### PR TITLE
nfd-master: tweak list options for NodeFeature informer

### PR DESCRIPTION
**Pull request by marquiz**
Fix cache syncing problems on big clusters with thousands of NodeFeature objects.

On the initial list (sync) the client-go cache reflector sets the ResourceVersion to "0" (instead of leaving it empty). This causes problems in the api server with (apiserver) logs like:

```
E writers.go:122] apiserver was unable to write a JSON response: http:
                  Handler timeout
E status.go:71] apiserver received an error that is not an
                metav1.Status: &errors.errorString{s:"http: Handler timeout"}:
                http: Handler timeout
```

On the nfd-master side we see corresponding log snippets like:

```
W reflector.go:547] failed to list *v1alpha1.NodeFeature: stream error
                    when reading response body, may be caused by closed
                    connection. Please retry. Original error: stream
                    error: stream ID 1521; INTERNAL_ERROR; received from
                    peer
I trace.go:236] "Reflector ListAndWatch" name:*** (***) (total time:
                61126ms): ---"Objects listed" error:stream error when
                reading response body, may be caused by closed
                connection. Please retry. Original error: stream
                error: stream ID 1521; INTERNAL_ERROR; received from
                peer 61126ms (***)
```

Decreasing the page size (opts.Limits) does not have any effect on the timeouts. However, setting ResourceVersion to an empty value seems to get the paging on its tracks, eliminating the timeouts.

TODO: investigate in Kubernetes upstream the root cause of the timeouts with ResourceVersion="0".

### CODE REVIEW

File: pkg/nfd-master/nfd-api-controller.go [hunk 1]
```diff
@@ -67,13 +68,23 @@ func newNfdController(config *restclient.Config, nfdApiControllerOptions nfdApiC
...
 	informerFactory := nfdinformers.NewSharedInformerFactory(nfdClient, nfdApiControllerOptions.ResyncPeriod)
 
 	// Add informer for NodeFeature objects
 	if !nfdApiControllerOptions.DisableNodeFeature {
-		featureInformer := informerFactory.Nfd().V1alpha1().NodeFeatures()
+		tweakListOpts := func(opts *metav1.ListOptions) {
+			// Tweak list opts on initial sync to avoid timeouts on the apiserver.
+			// NodeFeature objects are huge and the Kubernetes apiserver
+			// (v1.30) experiences http handler timeouts when the resource
+			// version is set to some non-empty value (TODO: find out why).
```
Comment (ahmetb) [inline pkg/nfd-master/nfd-api-controller.go:82]: I assume

* https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/1904-efficient-watch-resumption/README.md
* https://github.com/kubernetes/kubernetes/issues/123448

File: pkg/nfd-master/nfd-api-controller.go [hunk 2]
```diff
@@ -67,13 +68,23 @@ func newNfdController(config *restclient.Config, nfdApiControllerOptions nfdApiC
...
+			if opts.ResourceVersion == "0" {
+				opts.ResourceVersion = ""
+			}
+		}
+		featureInformer := nfdinformersv1alpha1.New(informerFactory, "", tweakListOpts).NodeFeatures()
```
Comment (ahmetb) [inline pkg/nfd-master/nfd-api-controller.go:87]: Q: doesn't nfd-gc need this tweak too?

File: pkg/nfd-master/nfd-api-controller.go (hunk 1 above)
Comment (marquiz) [reply pkg/nfd-master/nfd-api-controller.go:82]: This shouldn't happen as we basically only set rv="" on the initial LIST operation. When setting up the watches RV is always set >0

File: pkg/nfd-master/nfd-api-controller.go (hunk 2 above)
Comment (marquiz) [reply pkg/nfd-master/nfd-api-controller.go:87]: Indeed, needed there! I will update

**Review by PiotrProkop — COMMENTED** [review]
/lgtm

**Pull request merged**